commit 048d1267acddcaa03fe116e3a677e34ec2163e30
Author: 柳杨 <liu_yang42@dahuatech.com>
Date:   Fri Feb 27 16:57:11 2026 +0800

    修复位置删除错误

diff --git a/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java b/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
index 89fe453..5730c56 100644
--- a/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
+++ b/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
@@ -50,14 +50,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch
    * @param deletes the delete filter containing delete information
-   * @param rowStartPosInBatch the starting position of the row in the batch
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the size of the batch
    * @return the mapping array and the number of live rows, or {@code null} if nothing is deleted
    */
   public static Pair<int[], Integer> buildRowIdMapping(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     if (deletes == null) {
       return null;
@@ -70,7 +70,10 @@ public class ColumnarBatchUtil {
     int liveRowId = 0;
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
@@ -102,15 +105,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch.
    * @param deletes the delete filter containing information about which rows should be deleted.
-   * @param rowStartPosInBatch the starting position of the row in the batch, used to calculate the
-   *     absolute position of the rows in the context of the entire dataset.
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the number of rows in the current batch.
    * @return an array of boolean values to indicate if a row is deleted or not
    */
   public static boolean[] buildIsDeleted(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     boolean[] isDeleted = new boolean[batchSize];
 
@@ -123,7 +125,10 @@ public class ColumnarBatchUtil {
     ColumnarBatchRow row = new ColumnarBatchRow(columnVectors);
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
diff --git a/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java b/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
index ff30f29..45fdf57 100644
--- a/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
+++ b/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
@@ -181,12 +181,10 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
       }
 
       int numLiveRows = batch.numRows();
-      long rowStartPosInBatch =
-          rowPositionColumnIndex == -1 ? -1 : vectors[rowPositionColumnIndex].getLong(0);
 
       if (hasIsDeletedColumn) {
         boolean[] isDeleted =
-            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowPositionColumnIndex, numLiveRows);
         for (ColumnVector vector : vectors) {
           if (vector instanceof UpdatableDeletedColumnVector) {
             ((UpdatableDeletedColumnVector) vector).setValue(isDeleted);
@@ -194,7 +192,8 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
         }
       } else {
         Pair<int[], Integer> pair =
-            ColumnarBatchUtil.buildRowIdMapping(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildRowIdMapping(
+                vectors, deletes, rowPositionColumnIndex, numLiveRows);
         if (pair != null) {
           int[] rowIdMapping = pair.first();
           numLiveRows = pair.second();
diff --git a/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/data/parquet/vectorized/TestParquetPageSkippingVectorizedReads.java b/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/data/parquet/vectorized/TestParquetPageSkippingVectorizedReads.java
index 76b9781..9d38ec9 100644
--- a/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/data/parquet/vectorized/TestParquetPageSkippingVectorizedReads.java
+++ b/spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/data/parquet/vectorized/TestParquetPageSkippingVectorizedReads.java
@@ -209,7 +209,7 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
 
     // pos 180, 800 are deleted in file
     when(deletedRowPos.isDeleted(180L)).thenReturn(true);
-    when(deletedRowPos.isDeleted(800L)).thenReturn(true);
+    when(deletedRowPos.isDeleted(380L)).thenReturn(true);
     when(deleteFilter.deletedRowPositions()).thenReturn(deletedRowPos);
 
     Expression filter =
@@ -217,14 +217,13 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
             // page-1 in row group 0 -> row ranges: [100, 199]
             Expressions.and(
                 Expressions.greaterThanOrEqual("l", 150), Expressions.lessThan("l", 200)),
-            // firstRowIndex in row group 1 -> 619
-            // page-1, page-2 in row group 1 -> row ranges: [100, 299]
+            // page-3 in row group 0 -> row ranges: [300, 399]
             Expressions.and(
-                Expressions.greaterThanOrEqual("l", 750), Expressions.lessThan("l", 850)));
+                Expressions.greaterThanOrEqual("l", 350), Expressions.lessThan("l", 400)));
 
     List<Record> expected =
         selectRecords(
-            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(719, 800), Pair.of(801, 919));
+            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(300, 380), Pair.of(381, 400));
 
     readAndValidate(filter, expected, deleteFilter);
   }
diff --git a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java b/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
index 89fe453..5730c56 100644
--- a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
+++ b/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
@@ -50,14 +50,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch
    * @param deletes the delete filter containing delete information
-   * @param rowStartPosInBatch the starting position of the row in the batch
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the size of the batch
    * @return the mapping array and the number of live rows, or {@code null} if nothing is deleted
    */
   public static Pair<int[], Integer> buildRowIdMapping(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     if (deletes == null) {
       return null;
@@ -70,7 +70,10 @@ public class ColumnarBatchUtil {
     int liveRowId = 0;
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
@@ -102,15 +105,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch.
    * @param deletes the delete filter containing information about which rows should be deleted.
-   * @param rowStartPosInBatch the starting position of the row in the batch, used to calculate the
-   *     absolute position of the rows in the context of the entire dataset.
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the number of rows in the current batch.
    * @return an array of boolean values to indicate if a row is deleted or not
    */
   public static boolean[] buildIsDeleted(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     boolean[] isDeleted = new boolean[batchSize];
 
@@ -123,7 +125,10 @@ public class ColumnarBatchUtil {
     ColumnarBatchRow row = new ColumnarBatchRow(columnVectors);
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
diff --git a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java b/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
index ff30f29..45fdf57 100644
--- a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
+++ b/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
@@ -181,12 +181,10 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
       }
 
       int numLiveRows = batch.numRows();
-      long rowStartPosInBatch =
-          rowPositionColumnIndex == -1 ? -1 : vectors[rowPositionColumnIndex].getLong(0);
 
       if (hasIsDeletedColumn) {
         boolean[] isDeleted =
-            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowPositionColumnIndex, numLiveRows);
         for (ColumnVector vector : vectors) {
           if (vector instanceof UpdatableDeletedColumnVector) {
             ((UpdatableDeletedColumnVector) vector).setValue(isDeleted);
@@ -194,7 +192,8 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
         }
       } else {
         Pair<int[], Integer> pair =
-            ColumnarBatchUtil.buildRowIdMapping(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildRowIdMapping(
+                vectors, deletes, rowPositionColumnIndex, numLiveRows);
         if (pair != null) {
           int[] rowIdMapping = pair.first();
           numLiveRows = pair.second();
diff --git a/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java b/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
index dadbe3e..47bb844 100644
--- a/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
+++ b/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
@@ -297,6 +297,10 @@ public class TestColumnarBatchUtil {
     when(intVector.getInt(3)).thenReturn(43);
     when(intVector.getInt(4)).thenReturn(44);
 
+    for (int i = 0; i < 200; i++) {
+      when(intVector.getLong(i)).thenReturn((long) i);
+    }
+
     return new ColumnVector[] {intVector};
   }
 }
diff --git a/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java b/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
index 0163426..d31b48f 100644
--- a/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
+++ b/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
@@ -209,7 +209,7 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
 
     // pos 180, 800 are deleted in file
     when(deletedRowPos.isDeleted(180L)).thenReturn(true);
-    when(deletedRowPos.isDeleted(800L)).thenReturn(true);
+    when(deletedRowPos.isDeleted(380L)).thenReturn(true);
     when(deleteFilter.deletedRowPositions()).thenReturn(deletedRowPos);
 
     Expression filter =
@@ -217,14 +217,13 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
             // page-1 in row group 0 -> row ranges: [100, 199]
             Expressions.and(
                 Expressions.greaterThanOrEqual("l", 150), Expressions.lessThan("l", 200)),
-            // firstRowIndex in row group 1 -> 619
-            // page-1, page-2 in row group 1 -> row ranges: [100, 299]
+            // page-3 in row group 0 -> row ranges: [300, 399]
             Expressions.and(
-                Expressions.greaterThanOrEqual("l", 750), Expressions.lessThan("l", 850)));
+                Expressions.greaterThanOrEqual("l", 350), Expressions.lessThan("l", 400)));
 
     List<Record> expected =
         selectRecords(
-            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(719, 800), Pair.of(801, 919));
+            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(300, 380), Pair.of(381, 400));
 
     readAndValidate(filter, expected, deleteFilter);
   }
diff --git a/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java b/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
index 89fe453..5730c56 100644
--- a/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
+++ b/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
@@ -50,14 +50,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch
    * @param deletes the delete filter containing delete information
-   * @param rowStartPosInBatch the starting position of the row in the batch
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the size of the batch
    * @return the mapping array and the number of live rows, or {@code null} if nothing is deleted
    */
   public static Pair<int[], Integer> buildRowIdMapping(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     if (deletes == null) {
       return null;
@@ -70,7 +70,10 @@ public class ColumnarBatchUtil {
     int liveRowId = 0;
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
@@ -102,15 +105,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch.
    * @param deletes the delete filter containing information about which rows should be deleted.
-   * @param rowStartPosInBatch the starting position of the row in the batch, used to calculate the
-   *     absolute position of the rows in the context of the entire dataset.
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the number of rows in the current batch.
    * @return an array of boolean values to indicate if a row is deleted or not
    */
   public static boolean[] buildIsDeleted(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     boolean[] isDeleted = new boolean[batchSize];
 
@@ -123,7 +125,10 @@ public class ColumnarBatchUtil {
     ColumnarBatchRow row = new ColumnarBatchRow(columnVectors);
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
diff --git a/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java b/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
index ff30f29..45fdf57 100644
--- a/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
+++ b/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
@@ -181,12 +181,10 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
       }
 
       int numLiveRows = batch.numRows();
-      long rowStartPosInBatch =
-          rowPositionColumnIndex == -1 ? -1 : vectors[rowPositionColumnIndex].getLong(0);
 
       if (hasIsDeletedColumn) {
         boolean[] isDeleted =
-            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowPositionColumnIndex, numLiveRows);
         for (ColumnVector vector : vectors) {
           if (vector instanceof UpdatableDeletedColumnVector) {
             ((UpdatableDeletedColumnVector) vector).setValue(isDeleted);
@@ -194,7 +192,8 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
         }
       } else {
         Pair<int[], Integer> pair =
-            ColumnarBatchUtil.buildRowIdMapping(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildRowIdMapping(
+                vectors, deletes, rowPositionColumnIndex, numLiveRows);
         if (pair != null) {
           int[] rowIdMapping = pair.first();
           numLiveRows = pair.second();
diff --git a/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java b/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
index dadbe3e..47bb844 100644
--- a/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
+++ b/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
@@ -297,6 +297,10 @@ public class TestColumnarBatchUtil {
     when(intVector.getInt(3)).thenReturn(43);
     when(intVector.getInt(4)).thenReturn(44);
 
+    for (int i = 0; i < 200; i++) {
+      when(intVector.getLong(i)).thenReturn((long) i);
+    }
+
     return new ColumnVector[] {intVector};
   }
 }
diff --git a/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java b/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
index 4c9f2f3..7c86116 100644
--- a/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
+++ b/spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
@@ -209,7 +209,7 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
 
     // pos 180, 800 are deleted in file
     when(deletedRowPos.isDeleted(180L)).thenReturn(true);
-    when(deletedRowPos.isDeleted(800L)).thenReturn(true);
+    when(deletedRowPos.isDeleted(380L)).thenReturn(true);
     when(deleteFilter.deletedRowPositions()).thenReturn(deletedRowPos);
 
     Expression filter =
@@ -217,14 +217,13 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
             // page-1 in row group 0 -> row ranges: [100, 199]
             Expressions.and(
                 Expressions.greaterThanOrEqual("l", 150), Expressions.lessThan("l", 200)),
-            // firstRowIndex in row group 1 -> 613
-            // page-1, page-2 in row group 1 -> row ranges: [100, 299]
+            // page-3 in row group 0 -> row ranges: [300, 399]
             Expressions.and(
-                Expressions.greaterThanOrEqual("l", 750), Expressions.lessThan("l", 850)));
+                Expressions.greaterThanOrEqual("l", 350), Expressions.lessThan("l", 400)));
 
     List<Record> expected =
         selectRecords(
-            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(713, 800), Pair.of(801, 913));
+            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(300, 380), Pair.of(381, 400));
 
     readAndValidate(filter, expected, deleteFilter);
   }
diff --git a/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java b/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
index 89fe453..5730c56 100644
--- a/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
+++ b/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchUtil.java
@@ -50,14 +50,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch
    * @param deletes the delete filter containing delete information
-   * @param rowStartPosInBatch the starting position of the row in the batch
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the size of the batch
    * @return the mapping array and the number of live rows, or {@code null} if nothing is deleted
    */
   public static Pair<int[], Integer> buildRowIdMapping(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     if (deletes == null) {
       return null;
@@ -70,7 +70,10 @@ public class ColumnarBatchUtil {
     int liveRowId = 0;
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
@@ -102,15 +105,14 @@ public class ColumnarBatchUtil {
    *
    * @param columnVectors the array of column vectors for the batch.
    * @param deletes the delete filter containing information about which rows should be deleted.
-   * @param rowStartPosInBatch the starting position of the row in the batch, used to calculate the
-   *     absolute position of the rows in the context of the entire dataset.
+   * @param rowPositionColumnIndex the index of the row position column in the column vectors.
    * @param batchSize the number of rows in the current batch.
    * @return an array of boolean values to indicate if a row is deleted or not
    */
   public static boolean[] buildIsDeleted(
       ColumnVector[] columnVectors,
       DeleteFilter<InternalRow> deletes,
-      long rowStartPosInBatch,
+      int rowPositionColumnIndex,
       int batchSize) {
     boolean[] isDeleted = new boolean[batchSize];
 
@@ -123,7 +125,10 @@ public class ColumnarBatchUtil {
     ColumnarBatchRow row = new ColumnarBatchRow(columnVectors);
 
     for (int rowId = 0; rowId < batchSize; rowId++) {
-      long pos = rowStartPosInBatch + rowId;
+      long pos =
+          rowPositionColumnIndex == -1
+              ? rowId - 1
+              : columnVectors[rowPositionColumnIndex].getLong(rowId);
       row.rowId = rowId;
       if (isDeleted(pos, row, deletedPositions, eqDeleteFilter)) {
         deletes.incrementDeleteCount();
diff --git a/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java b/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
index ff30f29..45fdf57 100644
--- a/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
+++ b/spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
@@ -181,12 +181,10 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
       }
 
       int numLiveRows = batch.numRows();
-      long rowStartPosInBatch =
-          rowPositionColumnIndex == -1 ? -1 : vectors[rowPositionColumnIndex].getLong(0);
 
       if (hasIsDeletedColumn) {
         boolean[] isDeleted =
-            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildIsDeleted(vectors, deletes, rowPositionColumnIndex, numLiveRows);
         for (ColumnVector vector : vectors) {
           if (vector instanceof UpdatableDeletedColumnVector) {
             ((UpdatableDeletedColumnVector) vector).setValue(isDeleted);
@@ -194,7 +192,8 @@ abstract class BaseBatchReader<T extends ScanTask> extends BaseReader<ColumnarBa
         }
       } else {
         Pair<int[], Integer> pair =
-            ColumnarBatchUtil.buildRowIdMapping(vectors, deletes, rowStartPosInBatch, numLiveRows);
+            ColumnarBatchUtil.buildRowIdMapping(
+                vectors, deletes, rowPositionColumnIndex, numLiveRows);
         if (pair != null) {
           int[] rowIdMapping = pair.first();
           numLiveRows = pair.second();
diff --git a/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java b/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
index dadbe3e..47bb844 100644
--- a/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
+++ b/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/TestColumnarBatchUtil.java
@@ -297,6 +297,10 @@ public class TestColumnarBatchUtil {
     when(intVector.getInt(3)).thenReturn(43);
     when(intVector.getInt(4)).thenReturn(44);
 
+    for (int i = 0; i < 200; i++) {
+      when(intVector.getLong(i)).thenReturn((long) i);
+    }
+
     return new ColumnVector[] {intVector};
   }
 }
diff --git a/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java b/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
index 4c9f2f3..7c86116 100644
--- a/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
+++ b/spark/v4.1/spark/src/test/java/org/apache/iceberg/spark/data/vectorized/parquet/TestParquetPageSkippingVectorizedReads.java
@@ -209,7 +209,7 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
 
     // pos 180, 800 are deleted in file
     when(deletedRowPos.isDeleted(180L)).thenReturn(true);
-    when(deletedRowPos.isDeleted(800L)).thenReturn(true);
+    when(deletedRowPos.isDeleted(380L)).thenReturn(true);
     when(deleteFilter.deletedRowPositions()).thenReturn(deletedRowPos);
 
     Expression filter =
@@ -217,14 +217,13 @@ public class TestParquetPageSkippingVectorizedReads extends AvroDataTestBase {
             // page-1 in row group 0 -> row ranges: [100, 199]
             Expressions.and(
                 Expressions.greaterThanOrEqual("l", 150), Expressions.lessThan("l", 200)),
-            // firstRowIndex in row group 1 -> 613
-            // page-1, page-2 in row group 1 -> row ranges: [100, 299]
+            // page-3 in row group 0 -> row ranges: [300, 399]
             Expressions.and(
-                Expressions.greaterThanOrEqual("l", 750), Expressions.lessThan("l", 850)));
+                Expressions.greaterThanOrEqual("l", 350), Expressions.lessThan("l", 400)));
 
     List<Record> expected =
         selectRecords(
-            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(713, 800), Pair.of(801, 913));
+            allRecords, Pair.of(100, 180), Pair.of(181, 200), Pair.of(300, 380), Pair.of(381, 400));
 
     readAndValidate(filter, expected, deleteFilter);
   }
